rnc2rng/parser.py

```diff
@@ -549,11 +553,16 @@ def __init__(self, fn, src):
 else:
     str_types = str, bytes
 
+def detect_encoding(fn):
+    with open(fn, 'rb') as f:
+        bom = f.read(2)
+    return 'utf-16' if bom in (BOM_UTF16_BE, BOM_UTF16_LE) else 'utf-8'
+
 def parse(src=None, f=None):
     assert src is None or f is None
     if f is not None and isinstance(f, str_types):
         fn = f
-        with open(fn) as f:
+        with open(fn, encoding=detect_encoding(fn)) as f:
```

Comment:
Opening the file twice seems a little ugly. Since we do a full read anyway, we could always open the file in binary mode, then decode the contents according to the detected BOM, right?

Reply:
Agreed. That would disable the need to import `open` from `io` as well.